File: source/GyeongHo/boj/7576.py
import sys
input = sys.stdin.readline

m, n = map(int, input().split())
tomato = [list(map(int, input().split())) for _ in range(n)]
dy = (0, 0, 1, -1)
dx = (1, -1, 0, 0)

def bfs(q):
	MAX = -1
	while q:
		posY, posX = q.pop(0)
		for i in range(4):
			newY, newX = posY + dy[i], posX + dx[i]
			if 0 <= newY < n and 0 <= newX < m and tomato[newY][newX] == 0:
				tomato[newY][newX] = tomato[posY][posX] + 1
				q.append((newY, newX))
				MAX = max(MAX, tomato[newY][newX])
	for tot in tomato:
		if 0 in tot:
			return -1
	return MAX - 1

queue = []
for i in range(n):
	for j in range(m):
		if tomato[i][j] == 1:
			queue.append((i, j))

print(bfs(queue))

File: source/GyeongHo/boj/7576.cpp
#include <iostream>
#include <queue>
#include <utility>
using namespace std;

int m, n;
int tomato[1000][1000] = {0,};
int dy[] = {0, 0, 1, -1};
int dx[] = {1, -1, 0, 0};

queue<pair<int, int>> q;

int bfs()
{
	int MAX = 1;
	while(!q.empty()) {
		pair<int, int> pos = q.front();
		q.pop();
		for (int i = 0; i < 4; ++i) {
			int newY = pos.first + dy[i];
			int newX = pos.second + dx[i];
			if (newY >= 0 && newY < n && newX >= 0 && newX < m && tomato[newY][newX] == 0) {
				tomato[newY][newX] = tomato[pos.first][pos.second] + 1;
				q.push(make_pair(newY, newX));
				MAX = max(MAX, tomato[newY][newX]);
			}
		}
	}
	for (int i = 0; i < n; ++i)
		for (int j = 0; j < m; ++j)
			if (tomato[i][j] == 0)
				return -1;
	return MAX - 1;
}

int main()
{
	cin >> m >> n;
	for (int i = 0; i < n; ++i)
		for (int j = 0; j < m; ++j)
			cin >> tomato[i][j];

	for (int i = 0; i < n; ++i)
		for (int j = 0; j < m; ++j)
			if (tomato[i][j] == 1)
				q.push(make_pair(i, j));

	cout << bfs();
	return 0;
}
